### PR TITLE
Remove single token scaling factor access

### PR DESCRIPTION
**Pull request by nventuro**
The fact that we need to implement this function is one of the sources of confusion re. storage.

Turns out, we don't.

### CODE REVIEW

File: pkg/pool-stable-phantom/contracts/StablePoolStorage.sol [hunk 1]
```diff
@@ -220,16 +220,10 @@ abstract contract StablePoolStorage is BasePool {
...
-        if (token == _getToken3()) return _getScalingFactor3();
-        if (token == _getToken4()) return _getScalingFactor4();
-        if (token == _getToken5()) return _getScalingFactor5();
-        else {
-            _revert(Errors.INVALID_TOKEN);
-        }
+    function _scalingFactor(IERC20) internal view virtual override returns (uint256) {
+        // We never a single token's scaling factor: we aways process the entire array at once. Therefore, we don't
+        // bother providing an implementation for this.
+        _revert(Errors.UNIMPLEMENTED);
```
Comment (TomAFrench) [inline pkg/pool-stable-phantom/contracts/StablePoolStorage.sol:226]:
```suggestion
        // We never use a single token's scaling factor by itself, we always process the entire array at once.
        // Therefore we don't bother providing an implementation for this.
        _revert(Errors.UNIMPLEMENTED);
```

File: pkg/pool-stable-phantom/contracts/StablePoolStorage.sol [hunk 2]
```diff
@@ -220,16 +220,10 @@ abstract contract StablePoolStorage is BasePool {
...
-        if (token == _getToken1()) return _getScalingFactor1();
-        if (token == _getToken2()) return _getScalingFactor2();
-        if (token == _getToken3()) return _getScalingFactor3();
-        if (token == _getToken4()) return _getScalingFactor4();
-        if (token == _getToken5()) return _getScalingFactor5();
-        else {
-            _revert(Errors.INVALID_TOKEN);
-        }
+    function _scalingFactor(IERC20) internal view virtual override returns (uint256) {
+        // We never a single token's scaling factor: we aways process the entire array at once. Therefore, we don't
```
Comment (EndymionJkb) [inline pkg/pool-stable-phantom/contracts/StablePoolStorage.sol:224]:
```suggestion
        // We never need a single token's scaling factor: we aways process the entire array at once. Therefore, we don't
```

File: pkg/balancer-js/src/utils/errors.ts [hunk 3]
```diff
@@ -157,6 +157,7 @@ const balancerErrorCodes: Record<string, string> = {
   '601': 'FLASH_LOAN_FEE_PERCENTAGE_TOO_HIGH',
   '602': 'INSUFFICIENT_FLASH_LOAN_FEE_AMOUNT',
   '603': 'AUM_FEE_PERCENTAGE_TOO_HIGH',
+  '998': 'UNIMPLEMENTED',
```
Comment (EndymionJkb) [inline pkg/balancer-js/src/utils/errors.ts:160]: We have an "INVALID_OPERATION". Maybe "UNSUPPORTED_OPERATION"?

**Review by TomAFrench — APPROVED** [review]
Looks good to me. Only one comment change.

File: pkg/pool-stable-phantom/contracts/StablePoolStorage.sol [hunk 4]
```diff
@@ -220,16 +220,10 @@ abstract contract StablePoolStorage is BasePool {
...
-        if (token == _getToken3()) return _getScalingFactor3();
-        if (token == _getToken4()) return _getScalingFactor4();
-        if (token == _getToken5()) return _getScalingFactor5();
-        else {
-            _revert(Errors.INVALID_TOKEN);
-        }
+    function _scalingFactor(IERC20) internal view virtual override returns (uint256) {
+        // We never need a single token's scaling factor: we aways process the entire array at once. Therefore, we don't
+        // bother providing an implementation for this.
+        _revert(Errors.UNIMPLEMENTED);
```
Comment (nventuro) [inline pkg/pool-stable-phantom/contracts/StablePoolStorage.sol:226]:
```suggestion
        // We never use a single token's scaling factor by itself, we always process the entire array at once.
        // Therefore we don't bother providing an implementation for this.
        _revert(Errors.UNIMPLEMENTED);
```

File: pkg/balancer-js/src/utils/errors.ts (hunk 3 above)
Comment (nventuro) [reply pkg/balancer-js/src/utils/errors.ts:160]: Eh I'd say this is different, there's no 'operation'. It's just a piece from an abstract contract we choose to not implement.

**Pull request merged**